File: source/library/EnemyPart.js
//Copyright 2012, John Wilson, Brighton Sussex UK. Licensed under the BSD License. See licence.txt
var TURRETFULLSWING_DURATION	=	72 ;

function EnemyPart(enemy,width,height,ctype,tanim,offset,poly,radius,indes,weaponcollide,explosions)
{
	this.offset		=	offset ;
	this.anim		=	tanim ;
	this.enemy		=	enemy ;
	this.tx			=	offset.x ;
	this.ty			=	offset.y ;
	this.ctype		=	ctype ;
	this.direction	=	enemy.direction ;
	this.tangle		=	0 ;					// turret angle
	this.tphase		=	Math.random()*2*Math.PI ;
	this.time		=	0 ;
	var	 degaim		=	this.enemy.realpath ? 16 : 90
	this.aim		=	rad(degaim) ;

	var	turretFreq		=	(180/(degaim*TURRETFULLSWING_DURATION)) ;
	this.turretFreqCnst	=	2*Math.PI*turretFreq ;		// == 2*pi*freq
	
		
	var texFrames	=	tanim.getFrames() ;
	
//	alert("CTYPE "+this.ctype) ;
	
	this.frames		=	{} ;

	for (var idx = 0 ; idx < texFrames.length ; idx++)
	{
		this.frames[texFrames[idx]] = ImageResource.getResource(texFrames[idx]) ;
	}

}


EnemyPart.Create=function(enemy,width,height,ctype,tanim,offset,poly,radius,indes,weaponcollide,explosions)
{
	return new EnemyPart(enemy,width,height,ctype,tanim,offset,poly,radius,indes,weaponcollide,explosions) ;
}

function rad(deg)	{	return deg*Math.PI/180 ;	}
function deg(rad)	{	return rad*180/Math.PI ;	}

EnemyPart.prototype =
{
	constructor:		EnemyPart,

	Update:				function(dt)
						{
							this.time	=	this.time + dt ;
							
							if (this.anim)
							{
								this.anim.update(dt) ;
								
								
							}
							if (this.ctype && this.ctype == 'rotate') 
							{
								var ROTATESPEED	=	360 ;

								this.direction 	= 	this.direction + dt*ROTATESPEED ;
							}	

							if (this.ctype && this.ctype == 'turret') 
							{
								this.tangle 	= 	deg(this.aim*Math.sin(this.turretFreqCnst*this.time + this.tphase)) ;
							}	
							
						},
						
	GetFiringDirection:	function()
						{
							return (this.ctype && this.ctype == 'turret')  ? this.enemy.direction + this.tangle : this.direction ;
						},
	
 	Render:				function(rHelper)
						{
							if (this.anim)
							{
								var DEFAULTSCL	=	0.5 ;

								var scale		=	this.enemy.scale || DEFAULTSCL;
								var pos 		=	this.enemy.GetPosition() ;
								var direction	=	this.enemy.GetDirection() + this.direction ;
								
								var x 			=	pos._x, y = pos._y ;
								var img 		=	this.frames[this.anim.getRenderTexture()];
								var toffset		=	this.ctype == 'turret' ?  this.tangle : 0	
								rHelper.drawRawImage(img,x,y,direction+toffset,scale,scale) ;
							}
						},

	RenderDebug:		function(rHelper)
						{
						},

}

File: lua/EnemyPart.lua
-- Copyright 2012, John Wilson, Brighton Sussex UK. Licensed under the BSD License. See licence.txt

if not EnemyPart then

EnemyPart 			=	{}
EnemyPart.className	=	"EnemyPart"
EnemyPart.debug						=	false
local tmpV						=	Vector4.Create()
local sqrt						=	math.sqrt
local cos						=	math.cos
local sin						=	math.sin
local rad						=	math.rad

local kENEMY_MAXTURNANGLE			=	90.0
local kENEMY_ANGULAR_ACCELERATION	=	120.0


	-- EnemyPart.Create(width,height,tanim,poly,radius,indes,weaponcollide)

	function EnemyPart.Create(enemy,width,height,ctype,animation,offset,poly,radius,indesct,weaponcollide,explosions)

		local enemypart = {
			enemy		=	enemy,
			width		=	width	or 1,
			height		=	height	or 1,
			ctype		=	ctype,
			radius		=	radius  or 1,
			animation	=	animation,
			poly 		=	poly or 1,
			indes		=	indes,
			offset		=	offset or {x = 0, y = 0},
			explosions	=	explosions or {x = 0, y = 0},
			direction		=	0,
			weaponcollide	=	weaponcollide,
	
					currentAngularSpeed	=	0,
					currentAngularDir	=	0,
	
			className	=	EnemyPart.className,
		}
		assert(animation,'NIL ANIMATION '..enemy.name)
		setmetatable(enemypart,{ __index = EnemyPart })
		EnemyPart._Init(enemypart)
		return enemypart
	end


	function EnemyPart._Init(enemypart)
		enemypart.xoff		=	enemypart.offset.x
		enemypart.yoff		=	enemypart.offset.y
		enemypart.direction	=	enemypart.enemy.direction

		if (enemypart.animation.className == 'uvAnimation') then
			local loop			=	enemypart.animation.loop
			enemypart.sprite	=	enemypart.animation.sprite
			enemypart.animation	=	enemypart.animation.textureanimation
			TextureAnim.Play(enemypart.animation,true)
			
		else
			TextureAnim.Play(enemypart.animation,true)
		end
		
	end
	
	function EnemyPart.GetExplosions(enemypart)
		return enemypart.explosions
	end
	
	function EnemyPart.Render(enemypart,renderHelper,xpos,ypos,direction,scale)
		xpos	=	xpos + enemypart.xoff
		ypos	=	ypos + enemypart.yoff

		local tid = enemypart.animation:GetRenderTexture()
		local uvs	= tid.uvs

		if (uvs) then
		
			Sprite.SetUvs(enemypart.sprite,
				uvs.tlu,uvs.tlv,
				uvs.blu,uvs.blv,
				uvs.bru,uvs.brv,
				uvs.tru,uvs.trv
			)
		
			RenderHelper.DrawSprite(renderHelper,enemypart.sprite,xpos,ypos,scale,direction)
		
		else
			RenderHelper.DrawTexture(renderHelper,tid,xpos,ypos,direction ,scale)
		end
		
		EnemyPart.RenderDebug(enemypart,renderHelper,xpos,ypos,direction,scale)

	end
	

	
	function EnemyPart.RenderDebug(enemypart,renderHelper,xpos,ypos,direction,scale)
		
		if (EnemyPart.debug) then
			Polygon.SetScale(enemypart.poly,scale)
			Polygon.SetPosition(enemypart.poly,xpos,ypos)
			Polygon.Render(enemypart.poly,renderHelper,direction)
		end
		
	end


	function EnemyPart.Update(enemypart,dt)
		if (enemypart.animation) then
			--Logger.lprint("Enemy.Update Animation")
			enemypart.animation:Update(dt)
			--enemypart.follow	=	enemypart.enemy.follow
			--EnemyPart.Turn(enemypart,dt)
		end
	end
	
	function EnemyPart.GetVectorPosition(enemypart)
		return enemypart.enemy:GetVectorPosition()
	end
	


	function EnemyPart.Turn(enemypart,dt)

		local	diff_vec	=	Temp.v0 

		Vector4.Subtract(diff_vec,enemypart.follow:GetVectorPosition(), enemypart:GetVectorPosition())
		diff_vec:Normalise2()

		-- Work out target bearing
		
		local rot_tar = math.atan2(diff_vec:Y(), diff_vec:X())
		rot_tar = (math.deg(rot_tar)	+ 90)
		enemypart.targetbearing	=	rot_tar % 360

		enemypart.angdistance = ((enemypart.targetbearing - (enemypart.direction % 360)))

		enemypart.leftangdist = ((360-enemypart.targetbearing)	+ (enemypart.direction % 360)) % 360
		enemypart.rightangdist = ((enemypart.targetbearing + (360-(enemypart.direction % 360)))) % 360
		

		local	clockwiseDirection	=	(enemypart.rightangdist < enemypart.leftangdist) 
		
		
		if ( clockwiseDirection ~= enemypart.currentAngularDir ) then
			enemypart.currentAngularDir = clockwiseDirection
			enemypart.currentAngularSpeed = 0.0
		end

		enemypart.currentAngularSpeed = enemypart.currentAngularSpeed +(dt * kENEMY_ANGULAR_ACCELERATION*enemypart.turnrate)

		if ( enemypart.currentAngularSpeed > kENEMY_MAXTURNANGLE ) then
			enemypart.currentAngularSpeed = kENEMY_MAXTURNANGLE
		end
		
		local maxTurnAngle = enemypart.currentAngularSpeed * dt
		
		if(math.min(enemypart.rightangdist,enemypart.leftangdist)>maxTurnAngle) then
		
			if (clockwiseDirection) then
				enemypart.direction = (enemypart.direction + maxTurnAngle) % 360

			else
				-- go anti clockwise
				enemypart.direction = (enemypart.direction - maxTurnAngle) % 360
			end

		else
			enemypart.direction = enemypart.targetbearing
		end

	end
	
	function EnemyPart.CollidePoly(enemypart,xpos,ypos,px,py,direction,scale)
		return Polygon.IsPointInside2(enemypart.poly,xpos+enemypart.xoff,ypos+enemypart.yoff,px,py,direction,scale)
	end

end
